api/v1alpha1/warehouse_types.go

```diff
@@ -0,0 +1,171 @@
+package v1alpha1
+
+import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+// +kubebuilder:validation:Enum={SemVer,NewestBuild,Alphabetical,Digest}
+type ImageUpdateStrategy string
+
+const (
+	ImageUpdateStrategySemVer       ImageUpdateStrategy = "SemVer"
+	ImageUpdateStrategyNewestBuild  ImageUpdateStrategy = "NewestBuild"
+	ImageUpdateStrategyAlphabetical ImageUpdateStrategy = "Alphabetical"
+	ImageUpdateStrategyDigest       ImageUpdateStrategy = "Digest"
+)
+
+//+kubebuilder:object:root=true
+//+kubebuilder:subresource:status
+
+// Warehouse is a source of Freight.
+type Warehouse struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	// Spec describes sources of artifacts.
+	//
+	//+kubebuilder:validation:Required
+	Spec *WarehouseSpec `json:"spec"`
+	// Status describes the Warehouse's most recently observed state.
+	Status WarehouseStatus `json:"status,omitempty"`
+}
+
+func (w *Warehouse) GetStatus() *WarehouseStatus {
+	return &w.Status
+}
+
+// WarehouseSpec describes sources of versioned artifacts to be included in
+// Freight produced by this Warehouse.
+type WarehouseSpec struct {
+	// Subscriptions describes sources of artifacts to be included in Freight
+	// produced by this Warehouse.
+	//
+	//+kubebuilder:validation:Required
+	Subscriptions *RepoSubscriptions `json:"subscriptions"`
```

Comment:
Have we considered this alternative:

```golang
  Subscriptions []Subscription
```

with "oneof" like semantic for subscription: 

```golang
type Subscription struct {
	// Git describes subscriptions to Git repositories.
	Git *GitSubscription `json:"git,omitempty"`
	// Images describes subscriptions to container image repositories.
	Images *ImageSubscription `json:"images,omitempty"`
	// Charts describes subscriptions to Helm charts.
	Charts *ChartSubscription `json:"charts,omitempty"`
}
```

Reply:
We just spoke on Zoom, but for posterity, it is the way it is because "oneof" semantics are currently hard to enforce without doing that validation oneself in a webhook. I'm willing to add that validation in order to get slightly cleaner-looking YAML, but we agreed to do it in a follow-up.

I will open a separate issue for this so we don't lose track of it.

Reply:
#954